### PR TITLE
[Fix #269] use cached data

### LINKED ISSUE #269: Proposal: Use cached data directly, skip loading state when possible
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## What performance issue do we need to solve?

[[Slack discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1688580518417739)]

By default `withOnyx` works like this:
Wrap a component, and set a loading state, while Onyx gets the requested data from the storage (this is async)
When we have the data from the storage we update the state with the real data and set loading to `false`.
In general this is fine, however, this causes an extra render. And we want to avoid extra re-renders as much as possible to have a good performance.
We have the case that we have data already in the memory, which could be read synchronously at the first render of a onyx connected component. However, we always run though the async connect which causes the extra re-render, switching from loading state to the state with the final data.

## What is the impact of this on end-users?

- We often start with our components in a loading state, although we could show the content right away (improved UX)
- Opening screens often feels jumpy / not-smooth as we have a lot of re-renders

This is for example the current state of opening a chat screen in the new dot app:

https://github.com/Expensify/react-native-onyx/assets/16821682/2911ff16-8d85-4ff7-99b6-aeb88dfec56b

## List any benchmarks that show the severity of the issue

This is a react-devtoosl profiler recording of when the app opens. Roughly 66 commits and 1s until the last commit is made (the SidebarLinks becomes visible)

Screenshot 2023-06-28 at 12 36 48

Also you can measure the time it takes until the report screen is open and ready, which is in our testing around ~800ms.

## Proposed solution (if any)

Onyx internally already has a cache map. When a component connects, and the data is already in the cache map we return the data like this:
```js
function get(key) {
    // When we already have the value in cache - resolve right away
    if (cache.hasCacheForKey(key)) {
        return Promise.resolve(cache.getValue(key));
    }
```
This is good, however, it still runs through the async/callback and thus will cause an extra re-render.
I propose that we, in the constructor of the withOnyx wrapper, access the cache map directly and set the data immediately when we have them. We only set the loading state to true if we don’t have all the data right away.
This way, a withOnyx connected component will most of the times only render initially once, instead of twice.

(Note: once we do that we will notice that there are a few places in the `Onyx.js` code where we don't compare the data before sending it to the component. As a extension to this proposal we should wrap all those places with equal checks to avoid updating components with the same data twice).

## List any benchmarks after implementing the changes to show impacts of the proposed solution (if any)

Profiling again we get 44 commits, and just 0.6s for the initial render of SidebarLinks to finish!

Screenshot 2023-06-28 at 12 36 57

When it comes to opening the report screen we see 4x improvements, as it only takes 100-200ms to show the whole page:

https://github.com/Expensify/react-native-onyx/assets/16821682/8868bd11-3048-4625-a456-313631061ae7

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number: latest**
**Reproducible in staging?: yes**
**Reproducible in production?: yes**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**
**Issue reported by:**
**Slack conversation: https://expensify.slack.com/archives/C01GTK53T8Q/p1688580518417739**

### PR DESCRIPTION
**Pull request by hannojg**
### Details

This PR introduces support for `withOnyx` to load data from cache immediately and synchronously. This greatly increases the chance when rending a `withOnyx` wrapped component that it can render right away, instead of being blocked by a loading state at first.

Overview of the changes made: 

- **`tryGetCachedValue`**: New method in `Onyx.js` that is used by `withOnyx` to get values from cache
- **`withOnyx.js`**: In the constructor of the wrapper class try to load all data from cache. Only activate loading state, if not all data was available from cache.
- **`Onyx.js`**: 
  - Cleaned up the code a bit by combining certain statements
  - Added a few more equality checks before calling `setState`

### Related Issues

https://github.com/Expensify/react-native-onyx/issues/269

### Automated Tests

The tests have been modified to reflect the changes made. In particular a lot of the `waitForPromiseToResolve` calls were removed. They aren't needed anymore, as a component can render immediately, once we have the data in onyx, as we don't need to wait for the next tick.

### Linked PRs

/

### CODE REVIEW

**Review by kidroca — CHANGES_REQUESTED** [review]
Thank you for your efforts on this pull request. Overall, the changes look promising and I appreciate the thoughtfulness behind them. 

I've shared some specific comments for your consideration to improve the overall quality and maintainability of the code. 

One particular observation is that there seem to be some changes in this PR that aren't directly tied to the referenced issue. While these changes, which appear to be performance optimizations and refactoring, could indeed enhance our codebase, they also introduce extra complexity to the review process.

To streamline the review, I suggest we separate these optimizations and refactoring changes into a subsequent PR. This approach would allow us to maintain focus on addressing the primary objective of the current PR, and may even speed up its approval. 

Looking forward to your thoughts. Thanks again for your contributions!

File: lib/Onyx.js
Comment (Julesssss) [reply lib/Onyx.js:214]: +1 to the above comment change, thanks @kidroca!

File: lib/Onyx.js
Comment (Julesssss) [inline lib/Onyx.js:230]: What were we doing previously? Returning additional data from outside the selector?

**Review by Julesssss — COMMENTED** [review]
Great work, would love to see a bit more thorough comments for people like me who don't fully understand the logic.

File: lib/Onyx.js
Comment (hannojg) [reply lib/Onyx.js:215]: very good catch!

File: lib/Onyx.js
Comment (hannojg) [reply lib/Onyx.js]: Hey, good point, I updated the logic. When we didn't find any values for the collection key we shouldn't continue with the functions behaviour. So I added an early return statement for that.

File: lib/Onyx.js
Comment (hannojg) [reply lib/Onyx.js]: I removed the comment as it seems to be ambiguous. As this logic is new there is no previous behaviour, and the comment was just trying to say that we need to use the selector which is redundant

**Comment by hannojg** [comment]
Good for review again. I was also experimenting with removing the `if(loading) return null` logic, but that needs a bit more thorough testing.

**Comment by hannojg** [comment]
Sorry, fixed!

File: lib/Onyx.js
Comment (marcaaron) [inline lib/Onyx.js:211]:
```suggestion
            return;
```

File: lib/withOnyx.js
Comment (marcaaron) [inline lib/withOnyx.js:60]: nab, can avoid the spread with something like:

```js
cachedState.loading = loading;
this.state = cachedState;
```

File: lib/withOnyx.js
Comment (marcaaron) [inline lib/withOnyx.js:106]: Why this part?

```js
this.state[statePropertyName] !== val
```

Like, say we were able to set up the component in the constructor with all cache values. Loading is `false`. But if the value has changed we will start messing with `tempState` again? My expectation is that once we have all the values then we should be done with the other part of this method.

File: lib/withOnyx.js
Comment (marcaaron) [inline lib/withOnyx.js:53]: I think we only need to set the `cachedState` if we are still loading stuff.

File: lib/Onyx.js
Comment (marcaaron) [inline lib/Onyx.js:646]:
```suggestion
        // sendDataToConnection only gets called in the connect code.
```

File: lib/Onyx.js
Comment (marcaaron) [inline lib/Onyx.js:615]: while true, I'd maybe remove this as comments like this can become of date and confuse people.

File: lib/Onyx.js
Comment (marcaaron) [inline lib/Onyx.js:653]: Do we need a `deepEqual()` here? Wouldn't it be enough to see that it has the property name at all? It must have been set in the constructor? Or maybe I am missing some situation where these could actually be different.

File: lib/Onyx.js
Comment (marcaaron) [inline lib/Onyx.js:570]: Is this change related to the "init with cache data" fix?

File: lib/Onyx.js
Comment (marcaaron) [inline lib/Onyx.js:475]: Same here. Is this related or just something that seems more correct?

**Review by marcaaron — CHANGES_REQUESTED** [review]

File: lib/Onyx.js
Comment (arosiclair) [inline lib/Onyx.js:432]: Are there any performance implications on adding another deep equality check here?

Comment (marcaaron) [reply lib/Onyx.js:432]: Good point and - there might be. Guides e.g. can have 10,000 reports that would exist in a "collection"

File: lib/withOnyx.js
Comment (hannojg) [reply lib/withOnyx.js:104]: you are right, i double checked it, and this part isn't necessary anymore!

File: lib/withOnyx.js
Comment (hannojg) [reply lib/withOnyx.js:53]: Hm, not sure i understand you correctly. With the current changes we 

- always set all the cached state we could load for the keys to the component state 
- if not all keys have been retrievable from cache the loading value will be true

File: lib/Onyx.js
Comment (hannojg) [reply lib/Onyx.js:651]: Hm, you are right in a way. I think deepEqual is overkill, and a simple ref comparison should be enough. However we should do a check here. There is the situation where a prop provided to the component using withOnyx changes, which changes the key requested. In this case we connect the component again, and this code gets called again. However, at this point in time there will already be a state value. Thats why we can't return simply when the value seems to be in state>

Comment (hannojg) [reply lib/Onyx.js:651]: ah lol no, that wouldn't be enough, because we also have selectors, and their return value is always a different ref, even if the input data have been the same. Hmmm ...

Comment (hannojg) [reply lib/Onyx.js:656]: Okay, so i optimised it to use deepEqual for selectors only. Please check again

File: lib/Onyx.js
Comment (hannojg) [reply lib/Onyx.js:475]: Yeah again, you are right, thats just something thats belonging more to a different set of performance improvements. Will remove it from here and do a follow up issue/PR.

File: lib/Onyx.js
Comment (hannojg) [reply lib/Onyx.js]: removed it for now, lets discuss this in a separate issue / PR!

**Comment by hannojg** [comment]
Okay, lgtm now!

File: lib/Onyx.js
Comment (marcaaron) [reply lib/Onyx.js]: >prop provided to the component using withOnyx changes, which changes the key requested. In this case we connect the component again, and this code gets called again. However, at this point in time there will already be a state value. Thats why we can't return simply when the value seems to be in state.

Whoaaa! Did we see this happen? 🙃 

Let me make sure I understand the order of operations:

- Connect to a key with a prop (`withOnyx()` `constructor()` called)
- Component mounts and there is no value for something in cache so we try get the stored value (async operation)
- _Maybe_ we changed the prop influencing this key while we are getting the data from storage
- Component updates and we `disconnect()` (but don't cancel the reading from storage that is happening)
- We finish reading the first thing from storage and call `sendDataToConnection()`
- It will return early due to the logic on line 629

**Question:** What happens to the `loading` state in this scenario since we are never calling `setWithOnyxState()` and the value of `requiredKeysForInit` is never updated when the other key is disconnected? Does the component never render?

FWIW I am still not getting why we need the `deepEqual()`. Maybe clearer way to look at this is...

If we have the cached data then why are we calling `sendDataToConnection()` at all? Maybe we pass something to [this method](https://github.com/Expensify/react-native-onyx/blob/d4fc53cc3229808e9c1c542c13ce0089e5893d96/lib/withOnyx.js#L146C13-L157) and just skip it during the `connect()` method.

File: lib/withOnyx.js
Comment (marcaaron) [reply lib/withOnyx.js]: Sorry that was really unclear haha. 

I meant that setting up the `this.tempState` variable is not needed if we got the entire `cachedState`? Because we would never use it for anything.

(not sure why my brain said the other thing)

File: lib/withOnyx.js
Comment (marcaaron) [reply lib/withOnyx.js]: Great, let's remove it 😄

Comment (marcaaron) [reply lib/withOnyx.js]: wow it is gone GitHub trolled me 😂

File: lib/withOnyx.js
Comment (marcaaron) [inline lib/withOnyx.js:61]: Random (impossible to know) Expensify rule incoming, but we never put "TODO" in the code anywhere. something something it is better to make an issue so that we actually do whatever the "TODO" asks us to do.

Comment (marcaaron) [reply lib/withOnyx.js:61]: FWIW, I can't remember but this might _possibly_ be how we had it at some point. The data can be returned very fast and the component could takes a while to mount - then we end up setting state on an unmounted component (which throws the ugly warning). Honestly though - can't remember.

File: lib/withOnyx.js
Comment (marcaaron) [inline lib/withOnyx.js:74]: This comment looks like maybe it should be removed?

File: lib/withOnyx.js
Comment (marcaaron) [inline lib/withOnyx.js:105]: NAB, but maybe we should not call this at all in the case the cached data is there. This makes more sense to me vs. letting it get called but "cancelling" it. I also think the code we have here works no NAB.

**Review by marcaaron — COMMENTED** [review]
LGTM just need to remove the TODO comment and we are 🍒

File: lib/withOnyx.js
Comment (hannojg) [reply lib/withOnyx.js]: Oh, we continue using the tempState later on in `setStateWithOnyx`

**Review by arosiclair — APPROVED** [review]
LGTM other than Marc's comments

**Comment by j-piasecki** [comment]
I've tried your changes on the new architecture and they improve the performance greatly ❤️! One problem I've observed is that nesting reports sometimes results in the `ReportScreen` falling into an infinite render loop.

https://github.com/Expensify/react-native-onyx/assets/21055725/adabc7a7-f648-4ce0-b161-0e0d6d4b5944

I doubt it's directly tied to your changes, most probably it simply enabled a problem elsewhere but I wanted you to know just in case.

File: lib/withOnyx.js
Comment (marcaaron) [reply lib/withOnyx.js]: Only if `loading: true` ?

**Comment by marcaaron** [comment]
>I doubt it's directly tied to your changes, most probably it simply enabled a problem elsewhere but I wanted you to know just in case.

Oh, nice catch! Does it happen in the previous version? Sounds like something we should try to address when merging the update PR into `App`.

**Comment by hannojg** [comment]
Hey 👋 I think I'd like to test for that infinite render behaviour before we merge this, so we know that once we merge this PR it will work in the app.

Will try to reproduce and fix it today!

**Comment by j-piasecki** [comment]
> Will try to reproduce and fix it today!

Thanks! Here's a branch with the new arch setup: https://github.com/WoLewicki/App/tree/andrew-fabric. Let me know if you find anything, if not then I can look into it more tomorrow.

File: lib/withOnyx.js
Comment (hannojg) [reply lib/withOnyx.js:61]: yeah, I think this change would be a separate PR as well where we can look at and discuss all those cases (however, I might propose a bit of a architectural change to onyx in the following days which would handle this point as well)

File: lib/withOnyx.js
Comment (hannojg) [reply lib/withOnyx.js:105]: the condition is paired with the loading state, which is something the `withOnyx` is aware about, while `Onyx.js` isn't really. Hence I think it makes sense to leave the condition in this place 🤔

**Comment by hannojg** [comment]
Thank you! Checking now (would be great to get this PR merged today!)

File: lib/withOnyx.js
Comment (hannojg) [reply lib/withOnyx.js]: yeah, as `loading: true` is like "we didn't have all the data in cache and need to wait for storage or maybe even wait until a network request puts the data into onyx".

**Comment by j-piasecki** [comment]
@hannojg The render loop seems to be some weird interaction with `react-freeze` - setting `freeze` to `false` [here](https://github.com/Expensify/App/blob/88fb2ada376ce103224e30a028c48350db29f78d/src/libs/Navigation/FreezeWrapper.js#L46) seems to stop the render loop. Adding `initWithStoredValues: false` [here](https://github.com/Expensify/App/blob/main/src/pages/home/ReportScreen.js#L449-L451) also looks like it’s helping.

**Comment by hannojg** [comment]
Hey Thanks for the insights @j-piasecki! Will double check what's wrong with the usage of these!

**Comment by hannojg** [comment]
@j-piasecki and me just looked into the issue and agreed that the issue with the fabric architecture enabled is not a problem with the changes from this PR, but rather a problem with react-freeze. That means the issue isn't blocking this PR.

Green light! :shipit:

**Review by Julesssss — DISMISSED** [review]
Fantastic work. 

I would just suggest that we introduce a test specifically to protect us from a future regression that nulls the perf improvement. Specifically, it would be great if we can ensure [this condition](https://github.com/Expensify/react-native-onyx/pull/280/files#diff-364fd7349a1af7b52091427c5a0734da8ff56057fbc7fe2be10734a06c97264cR52-R53) is always met.

**Comment by hannojg** [comment]
@Julesssss I already added a test earlier that checks if the component is rendered immediately with the data from cache. I extended this test to make sure the component doesn't re-render with the same data (e.g. after the connection was created and `Onyx.js` tries to send the data with `sendDataToConnection`)

**Review by Julesssss — DISMISSED** [review]

File: tests/unit/withOnyxTest.js
Comment (mountiny) [inline tests/unit/withOnyxTest.js:277]: I know this was there but
```suggestion
            // When a single item in the collection is updated with mergeCollection()
```

**Review by mountiny — APPROVED** [review]
Thanks! I think would be great to get approvals of @marcaaron and @tgolen before merging

File: lib/Onyx.js
Comment (tgolen) [inline lib/Onyx.js:213]: NAB 
```suggestion
        if (_.size(values) === 0) {
```

File: lib/withOnyx.js
Comment (tgolen) [inline lib/withOnyx.js:50]: NAB: This could be a `_.reduce()` instead.

File: lib/withOnyx.js
Comment (tgolen) [inline lib/withOnyx.js:53]:
```suggestion
                cachedState.loading = _.size(cachedState) < requiredKeysForInit.length;
```

File: lib/withOnyx.js
Comment (tgolen) [inline lib/withOnyx.js:104]: Could you please update this comment to answer the question "Why do we have to check whether the data has changed or not before setting the state?" Right now, the comment only explains what is happening and not _why_ it is happening, and I'm not entirely sure about the reason.

File: lib/withOnyx.js
Comment (tgolen) [inline lib/withOnyx.js:129]:
```suggestion
                // The state is set here manually, instead of using setState because setState is an async operation, meaning it might execute on the next task,
                // or at a later point in the microtask queue. That can lead to a race condition where
                // setWithOnyxState is called before the state is actually set. This results in unreliable behavior when checking the loading state and has been mainly observed on fabric.
```

**Review by tgolen — CHANGES_REQUESTED** [review]
Just a few small things I noticed that can be buttoned up a little.

File: lib/withOnyx.js
Comment (hannojg) [reply lib/withOnyx.js]: I updated it, I hope its more obvious now?

**Review by tgolen — DISMISSED** [review]

**Review by luacmartins — COMMENTED** [review]
Looking good. Left a couple of small comments

File: lib/Onyx.js
Comment (marcaaron) [reply lib/Onyx.js:196]: Thinking more on it. If I delete a key by setting it to `null` - it should not go back to the default key state (unless I call `clear()` or `init()` again). But maybe that behavior is disputed.

**Review by marcaaron — CHANGES_REQUESTED** [review]
Looks good, but I have some concern about using the `defaultKeyStates` as a cache value fallback. I think this behavior is wrong.

File: lib/Onyx.js
Comment (hannojg) [reply lib/Onyx.js:196]: I think you have a point here. The defaultKeyStates should be set to the cache on init. Thus it's introducing a wrong behaviour when falling back to the default here. Will change that, thank you!

File: lib/withOnyx.js
Comment (mountiny) [inline lib/withOnyx.js:56]: I am not sure why do we set the tempState to the same cachedState as the normal state. I might be missing something but could you help me understand this? Should we update the comment above?

**Review by mountiny — APPROVED** [review]
Thanks for pushing this forwards!

**Review by luacmartins — APPROVED** [review]

**Review by Julesssss — APPROVED** [review]

**Review by tgolen — APPROVED** [review]

**Comment by mountiny** [comment]
Per the slack discussion we got enough approvals here and we want to get this change out to test how it improves performance so I am going to merge this https://expensify.slack.com/archives/C035J5C9FAP/p1690473404613159?thread_ts=1690290738.145189&cid=C035J5C9FAP

File: lib/withOnyx.js
Comment (marcaaron) [reply lib/withOnyx.js:56]: The intention here is that `tempState` is still used in the same way as it was before. 

If we did not get all the values from the cache then the values are held in the `tempState` until all are "loaded". This way we avoid calling `setState()` each time a value comes back from storage. That was the original optimization added here and is a _good thing_ because otherwise you may have like several calls to setState() happening in a kind of burst before the first render (at least that's what we saw and why we changed how this works).

Though the question did make me wonder about something... if `this.tempState` and `this.state` are initialized with the same object reference then will we be mutating the state directly on this line:

https://github.com/Expensify/react-native-onyx/blob/80c6f001e27e798ed56cc4f3a39ee27c5efae375/lib/withOnyx.js#L117

Seems like that would be undesirable in the case where you did not initialize all the values in the constructor.

**Comment by marcaaron** [comment]
haven't tested the theory, but this might be a problem -> https://github.com/Expensify/react-native-onyx/pull/280#discussion_r1276988396

File: lib/withOnyx.js
Comment (mountiny) [reply lib/withOnyx.js:56]: Thanks for the write up. @hannojg what do you think of the concern Marc raised above?

**Pull request merged**
